Add performJITMemcpyAtomic and simplify jit copying code [2].

justinmichaud opened the pull request:
#### ebda8e26827a74360113aa00e52763aac48c9b58
<pre>
Add performJITMemcpyAtomic and simplify jit copying code [2].
<a href="https://bugs.webkit.org/show_bug.cgi?id=297235">https://bugs.webkit.org/show_bug.cgi?id=297235</a>

Reviewed by Keith Miller.

The eventual goal of this series of patches is to make clear, explicit and separate
entrypoints into code that eventually copies to the jit region or an assembler buffer,
to clearly separate patching that must be atomic and patching that does not.

This is important to ensure that every place relying on atomic copying behaviour
is explicitly documented, so that ARMv7 can ensure the correct alignment of these regions.

To start, we carefuly thread through a new atomic flag, but we avoid changing behavior yet.

The only change this patch should make is that sometimes we can avoid a buffer allocation
when linking.

Canonical link: <a href="https://commits.webkit.org/298849@main">https://commits.webkit.org/298849@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f437e61233c7b6b3badd109114032b09887d560e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116624 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36288 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26849 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122697 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67196 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36986 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44877 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88573 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/43006 "Too many flaky failures: fast/canvas/offscreen-no-script-context-crash.html, fast/css/vertical-align-rem-on-root.html, http/tests/permissions/storage-access-permissions-query.html, imported/w3c/web-platform-tests/event-timing/contextmenu.html, imported/w3c/web-platform-tests/trusted-types/TrustedTypePolicyFactory-getAttributeType-event-handler-content-attributes.tentative.html, imported/w3c/web-platform-tests/trusted-types/set-event-handlers-content-attributes.tentative.html, js/dom/string-prototype-properties.html, storage/indexeddb/cursor-value-private.html, storage/indexeddb/intversion-pending-version-changes-ascending-private.html, webgl/2.0.0/conformance2/textures/canvas_sub_rectangle/tex-2d-rgb16f-rgb-half_float.html, webgl/2.0.0/conformance2/textures/image/tex-2d-rgba4-rgba-unsigned_byte.html, webgl/2.0.0/conformance2/textures/image_bitmap_from_canvas/tex-2d-r8ui-red_integer-unsigned_byte.html, webgl/2.0.0/conformance2/textures/image_bitmap_from_canvas/tex-3d-rgb8-rgb-unsigned_byte.html, webgl/2.0.0/conformance2/textures/image_bitmap_from_image_data/tex-3d-rgba4-rgba-unsigned_byte.html, webgl/2.0.0/conformance2/textures/webgl_canvas/tex-2d-r11f_g11f_b10f-rgb-unsigned_int_10f_11f_11f_rev.html, webgl/2.0.y/conformance/ogles/GL/degrees/degrees_001_to_006.html, webgl/2.0.y/conformance2/query/occlusion-query.html, webgl/2.0.y/conformance2/transform_feedback/transform_feedback.html (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119573 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29500 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104622 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69039 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28567 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22728 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66365 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/108738 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98882 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22887 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125834 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/115152 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43523 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32697 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97244 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43887 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100824 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97037 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24754 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42358 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20286 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39485 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43409 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49004 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/143850 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42875 "Built successfully") | | [💥 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37032 "An unexpected error occured. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46215 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44581 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/697fb6f5b9ff7caeae7476b5c078c4608793c8c6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49231

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/0f9e283b513c5a8868eaa32a0e5808b6ad9634ab. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49231

justinmichaud committed:
Add performJITMemcpyAtomic and simplify jit copying code [2].
https://bugs.webkit.org/show_bug.cgi?id=297235

Reviewed by Keith Miller.

The eventual goal of this series of patches is to make clear, explicit and separate
entrypoints into code that eventually copies to the jit region or an assembler buffer,
to clearly separate patching that must be atomic and patching that does not.

This is important to ensure that every place relying on atomic copying behaviour
is explicitly documented, so that ARMv7 can ensure the correct alignment of these regions.

To start, we carefuly thread through a new atomic flag, but we avoid changing behavior yet.

The only change this patch should make is that sometimes we can avoid a buffer allocation
when linking.

Canonical link: https://commits.webkit.org/298849@main
